theano/tensor/opt.py

```diff
@@ -618,6 +618,26 @@ def local_dimshuffle_lift(node):
             copy_stack_trace(node.outputs[0], ret)
             return [ret]
 
+    # remove useless dimshuffle in general
+    # covers two types of useless dimshuffle:
+    #   1 - dimshuffle all dimensions in order
+    #   2 - dimshuffle a broadcastable dimension
```

Comment:
This looks correct, thanks!
I think it would be better if we also did it for dimshuffles caused by merging, above. We can refactor that into one function that you would call in both places.